internal/mvcc/backend/batch_tx.go

```diff
@@ -231,6 +232,13 @@ func (t *batchTxBuffered) CommitAndStop() {
 }
 
 func (t *batchTxBuffered) commit(stop bool) {
+	// no need to reset read tx if there is no write.
+	// call commit to update stats.
+	if t.batchTx.pending == 0 && !stop {
```

Comment:
if this approach isn't working (deadlocking in CI?), it might be easier to move the `pending == 0` paths to backend.run() and avoid calling commit() there...

Reply:
CI panics since the tx initialization code goes through here. I am still playing with the code to see what is the easier. Probably I should try what you just suggested.